lib/mini_racer/shared.rb

```diff
@@ -0,0 +1,380 @@
+
+      # defined in the C class
+      init_with_snapshot(snapshot)
+    end
+  end
+
+  class Platform
+    class << self
+      def set_flags!(*args, **kwargs)
+        flags_to_strings([args, kwargs]).each do |flag|
+          # defined in the C class
+          set_flag_as_str!(flag)
+        end
+      end
+
+      private
+
+      def flags_to_strings(flags)
+        flags.flatten.map { |flag| flag_to_string(flag) }.flatten
+      end
+
+      # normalize flags to strings, and adds leading dashes if needed
+      def flag_to_string(flag)
+        if flag.is_a?(Hash)
+          flag.map do |key, value|
+            "#{flag_to_string(key)} #{value}"
+          end
+        else
+          str = flag.to_s
+          str = "--#{str}" unless str.start_with?('--')
+          str
+        end
+      end
+    end
+  end
+
+  # eval is defined in the C class
+  class Context
+
+    class ExternalFunction
+      def initialize(name, callback, parent)
+        unless String === name
+          raise ArgumentError, "parent_object must be a String"
+        end
+        parent_object, _ , @name = name.rpartition(".")
+        @callback = callback
+        @parent = parent
+        @parent_object_eval = nil
+        @parent_object = nil
+
+        unless parent_object.empty?
+          @parent_object = parent_object
+
+          @parent_object_eval = ""
+          prev = ""
+          first = true
+          parent_object.split(".").each do |obj|
+            prev << obj
+            if first
+              @parent_object_eval << "if (typeof #{prev} !== 'object' || typeof #{prev} !== 'function') { #{prev} = {} };\n"
+            else
+              @parent_object_eval << "#{prev} = #{prev} || {};\n"
+            end
+            prev << "."
+            first = false
+          end
+          @parent_object_eval << "#{parent_object};"
+        end
+        notify_v8
+      end
+    end
+
+    def initialize(max_memory: nil, timeout: nil, isolate: nil, ensure_gc_after_idle: nil, snapshot: nil, marshal_stack_depth: nil)
+      options ||= {}
+
+      check_init_options!(isolate: isolate, snapshot: snapshot, max_memory: max_memory, marshal_stack_depth: marshal_stack_depth, ensure_gc_after_idle: ensure_gc_after_idle, timeout: timeout)
+
+      @functions = {}
+      @timeout = nil
+      @max_memory = nil
+      @current_exception = nil
+      @timeout = timeout
+      @max_memory = max_memory
+      @marshal_stack_depth = marshal_stack_depth
+
+      # false signals it should be fetched if requested
+      @isolate = isolate || false
+
+      @ensure_gc_after_idle = ensure_gc_after_idle
+
+      if @ensure_gc_after_idle
+        @last_eval = nil
+        @ensure_gc_thread = nil
+        @ensure_gc_mutex = Mutex.new
+      end
+
+      @disposed = false
+
+      @callback_mutex = Mutex.new
+      @callback_running = false
+      @thread_raise_called = false
+      @eval_thread = nil
+
+      # defined in the C class
+      init_unsafe(isolate, snapshot)
```

Comment:
init_unsafe no longer exists

Reply:
It does, it's defined in `lib/mini_racer/truffleruby.rb`.